File: elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(1, _), do: 0
  def to(limit, factors) do
    do_to(limit - 1, factors)
  end

  defp do_to(1, _), do: 0
  defp do_to(item, factors) do
    case Enum.any?(factors, &(rem(item, &1) == 0)) do
      true -> item + do_to(item - 1, factors)
      _    -> do_to(item - 1, factors)
    end
  end
end

File: javascript/sum-of-multiples/sum-of-multiples.js
var SumOfMultiples = function(factors) {
  if (!(this instanceof SumOfMultiples)) {
    return new SumOfMultiples(factors);
  }
  this.factors = factors;
};

SumOfMultiples.prototype.to = function(limit) {
  let out = 0;
  for (var x=1; x<limit; x++) {
    if (this.factors.some((f) => x % f == 0)) {
      out += x;
    }
  }
  return out;
};

module.exports = SumOfMultiples;
